### PR TITLE
Closeness centrality

### PR DESCRIPTION
**Pull request by bmckown**
This is the next pr in line for adding more algorithms to cassovary.  This one addresses closeness centrality and was adopted from and checked against this [library](http://networkx.github.io/documentation/networkx-1.9.1/_modules/networkx/algorithms/centrality/closeness.html#closeness_centrality).

AbstractCentrality was slightly refactored to reduce code redundancy that become apparent after adding an additional centrality algorithm.  Tests were also added.

### CODE REVIEW

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package com.twitter.cassovary.algorithms.centrality
+
+import com.twitter.cassovary.graph.{GraphDir, BreadthFirstTraverser, Walk, DirectedGraph}
+import scala.collection.immutable
+
+/**
+ * Calculate the closeness centrality of {@code graph}.
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: Even though the current code is littered with {@code ...} we don't use it anymore. Prefer `graph`

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * @param normalize Normalize by the ratio of number of connections to other nodes
+ *                  and the number of nodes in the graph
+ */
+class ClosenessCentrality(graph: DirectedGraph, normalize: Boolean = true) extends AbstractCentrality(graph) {
+  /**
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: Not a real useful comment over and beyond one in abstract class. I'd delete it.

**Comment by bmckown** [comment]
Docs have been updated.

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala [hunk 3]
```diff
@@ -17,7 +17,7 @@ import collection.immutable
 import com.twitter.cassovary.graph.{Node, DirectedGraph}
 
 /**
- * Centrality algorithm abstraction that requires a directed graph {@code graph} be passed upon
+ * Centrality algorithm abstraction that requires a directed graph (<code>graph</code>) be passed upon
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala]: i meant to say backticks graph backticks (markdown may have caused it to appear as something else).

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ */
+class ClosenessCentrality(graph: DirectedGraph, normalize: Boolean = true) extends AbstractCentrality(graph) {
+
+  override def recalculate(): immutable.IndexedSeq[Double] = {
+    graph foreach { node =>
+      val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(node.id), Walk.Limits())
+      var sum = 0.0
+      var neighborCount = 0.0
+
+      bfs foreach { n =>
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: make this functional style.
val (sum, neighborCount) = bfs.foldLeft((0,0)) { case ((partialSum, partialCount),a) => (partialSum+a, partialSum+1) } }

Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: substitute 'a' above with bfs.depth(n.id).get

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      }
+
+      centralityValues(node.id) = {
+        if (sum > 0 && graph.nodeCount > 1) {
+          val numerator = neighborCount - 1
+          val denom = if (normalize) (graph.nodeCount - 1) / (neighborCount - 1) else 1
+          numerator / (denom * sum)
+        }
+        else
+          0.0
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: put 0.0 in the same line as else

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(node.id), Walk.Limits())
+      var sum = 0.0
+      var neighborCount = 0.0
+
+      bfs foreach { n =>
+        sum += bfs.depth(n.id).get
+        neighborCount += 1
+      }
+
+      centralityValues(node.id) = {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: don't need the extra '{' and pull up next line (if....) to this line

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        if (sum > 0 && graph.nodeCount > 1) {
+          val numerator = neighborCount - 1
+          val denom = if (normalize) (graph.nodeCount - 1) / (neighborCount - 1) else 1
+          numerator / (denom * sum)
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: just confirming that it is either (neighborCount - 1) / sum or (neighborCount-1)^2 / ((nodeCount - 1) * sum) ?

(also if you use the above foldLeft, will need to use sum.toDouble and neighborCount.toDouble to calculate denom)

btw, looks like in neighborCount you are really counting reachable nodes, perhaps change variable name to reachableCount to be clearer?

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        else
+          0.0
+      }
+    }
+    centralityValues.toIndexedSeq
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: looks like this is common pattern. You might want to refactor AbstractCentrality as:

private val centralityValues = new Array....

protected def _recalc(centr: Array[Int]): Unit

def recalculate(): immutable.indexedSeq = {
_recalc(centralityValues)
centralityValues.toIndexedSeq
}

and implement _recalc() in every subclass.

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala (hunk 3 above)
Comment (bmckown) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala]: Ha.  Yep.  Markdown definitely made it look exactly like < code > graph < /code >

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala (hunk 7 above)
Comment (bmckown) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: just confirming that it is either (neighborCount - 1) / sum or (neighborCount-1)^2 / ((nodeCount - 1) * sum) ?

You are correct.

**Comment by bmckown** [comment]
Updated.  Let me know what you think!

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala [hunk 9]
```diff
@@ -25,6 +25,7 @@ abstract class AbstractCentrality(graph: DirectedGraph) extends Centrality {
 
   protected val centralityValues = new Array[Double](graph.maxNodeId + 1)
 
+  recalculate()
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/AbstractCentrality.scala]: you can just call _recalc() here as you don't care about the return value

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import com.twitter.cassovary.graph.{GraphDir, BreadthFirstTraverser, Walk, DirectedGraph}
+
+/**
+ * Calculate the closeness centrality of ``graph``
+ * @param normalize Normalize by the ratio of number of connections to other nodes
+ *                  and the number of nodes in the graph
+ */
+class ClosenessCentrality(graph: DirectedGraph, normalize: Boolean = true) extends AbstractCentrality(graph) {
+
+  override def _recalc(): Unit = {
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: don't need 'override' keyword

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    graph foreach { node =>
+      val bfs = new BreadthFirstTraverser(graph, GraphDir.OutDir, Seq(node.id), Walk.Limits())
+
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: delete empty line

File: cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      val (sum, reachableNodes) = bfs.foldLeft((0.0, 0.0)) {
+        case ((partialSum, partialCount), a) => (partialSum + bfs.depth(a.id).get, partialCount + 1)
+      }
+
```
Comment (pankajgupta) [inline cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/centrality/ClosenessCentrality.scala]: delete empty line

**Comment by pankajgupta** [comment]
LGTM pending above minor comments. Please do a squash commit before we merge it.

**Comment by bmckown** [comment]
Do you mind if I push the change to this PR or would you prefer a fresh PR to be submitted?

**Comment by pankajgupta** [comment]
same PR is good

**Comment by bmckown** [comment]
Squashed.

**Pull request merged**